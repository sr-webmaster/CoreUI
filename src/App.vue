<template>
  <v-app id="inspire">
    <component :is="layout">
      <slot />
    </component>
  </v-app>
</template>

<script>
import { mapActions } from 'vuex'

export default {
  computed: {
    layout () {
      return 'layout-' + (this.$route.meta.layout || 'default')
    }
  },
  beforeMount () {
    this.getCurrentUser()
  },
  methods: {
    ...mapActions('currentUser', ['getCurrentUser'])
  }
}
</script>

<style scoped>

</style>
