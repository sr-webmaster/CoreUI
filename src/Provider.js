import currentUser from './store/modules/currentUser'
import navigation from './store/modules/navigation'
import navigationAdmin from './store/modules/navigationAdmin'
import userNotifications from './store/modules/userNotifications'
import { version, name } from '../package.json'

export default () => {
  return {
    name,
    layouts: require.context('./layouts', false, /\.vue$/),
    store: {
      currentUser,
      navigation,
      navigationAdmin,
      userNotifications
    },
    version
  }
}
