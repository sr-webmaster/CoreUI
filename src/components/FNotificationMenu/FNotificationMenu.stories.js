import { MAIN } from '../../../.storybook/categories'
import FNotificationMenu from './FNotificationMenu.vue'
import MockAdapter from 'axios-mock-adapter'
import axios from 'axios'
import Vue from 'vue'
import Vuex from 'vuex'
import UserNotifications from '../../store/modules/userNotifications'

const mock = new MockAdapter(axios)
Vue.use(Vuex)

const getTitle = () => {
  return `${MAIN}|beta/Notification Menu`
}
export default {
  title: getTitle(),
  id: 'FNotificationMenu'
}

const serverData = [
  {
    id: 1,
    icon: 'fas fa-file-alt',
    created_at: 'Mar 15, 10:32 AM',
    event: 'John Smith shared',
    title: 'Retail Industry Report Q4 2018',
    route: '/content/100',
    msg: ''
  },
  {
    id: 2,
    icon: 'fas fa-file-alt',
    created_at: 'Mar 15, 10:32 AM',
    event: 'John Smith shared',
    title: '8 Important Questions Answered by Proposed Regulations on Qualified Opportunity Funds',
    route: '/content/1',
    msg: 'Hey, John. This is really interesting. Please check this out.'
  }
]

mock
  .onGet('/users/1/notifications').reply(200, {
    data: serverData
  })
  .onAny('/users/1/notifications/1').reply(200, {
    data: serverData[0]
  })
  .onAny('/users/1/notifications/2').reply(200, {
    data: serverData[1]
  })

const notificationsStore = new Vuex.Store({
  getters: {
    currentUser () {
      return { id: 1 }
    }
  },

  modules: {
    userNotifications: UserNotifications({})
  }
})

/**
 * https://storybook.js.org/docs/guides/guide-vue/#step-4-write-your-stories
 */
export const defaultStory = () => ({
  components: { FNotificationMenu },
  store: notificationsStore,
  template: `
        <v-container grid-list-xl>
            <v-layout column>
              <v-flex>
                <v-alert
                  :value="true"
                  color="warning"
                  icon="priority_high"
                >
                  This component is currently in Beta. As such it can change without warning or semver Major version updates
                </v-alert>
              </v-flex>
              <v-flex>
                <f-notification-menu style="width: 320px;" />
                <v-btn @click="reset">Reset</v-btn>
              </v-flex>
            </v-layout>
        </v-container>
      `,
  methods: {
    reset () {
      this.$store.dispatch('userNotifications/getItems', { params: { userId: 1 } })
    }
  }
})
defaultStory.story = {
  name: 'defaults (uses Store)'
}
