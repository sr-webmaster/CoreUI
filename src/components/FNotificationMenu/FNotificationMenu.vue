<template>
  <v-card>
    <v-list>
      <v-list-tile
        v-if="notifications.length"
        class="clear-tile"
      >
        <v-list-tile-avatar />
        <v-list-tile-content />
        <v-list-tile-action>
          <v-btn
            icon
            title="Dismiss all notifications..."
            @click.prevent="dismissAll"
          >
            <v-icon class="grey--text">
              clear_all
            </v-icon>
          </v-btn>
        </v-list-tile-action>
      </v-list-tile>

      <template v-for="notification in notifications">
        <v-divider :key="'divider-' + notification.id" />
        <v-list-tile
          :key="notification.id"
          avatar
          :href="notification.route"
        >
          <v-list-tile-avatar>
            <v-icon class="secondary white--text">
              {{ notification.icon }}
            </v-icon>
          </v-list-tile-avatar>

          <v-list-tile-content>
            <v-list-tile-sub-title
              class="notification-headline"
              :title="headline(notification)"
            >
              {{ headline(notification) }}
            </v-list-tile-sub-title>
            <v-list-tile-title class="notification-label secondary--text">
              {{ notification.title }}
            </v-list-tile-title>
            <v-list-tile-sub-title class="notification-msg">
              {{ notification.msg }}
            </v-list-tile-sub-title>
          </v-list-tile-content>

          <v-list-tile-action>
            <v-btn
              icon
              title="Dismiss notification..."
              @click.prevent="dismiss(notification)"
            >
              <v-icon class="grey--text">
                fas fa-times
              </v-icon>
            </v-btn>
          </v-list-tile-action>
        </v-list-tile>
      </template>
    </v-list>
  </v-card>
</template>

<script>
import { mapGetters } from 'vuex'

/**
 * Please note that without Semver Major change we will introduce a breaking change to this component soon.
 *  **Do not use directly** Use only indirectly through the Core UI Layouts
 */
export default {
  props: {
    interval: {
      type: Number,
      default: 60000
    }
  },
  data: () => ({
    fetchInterval: null
  }),
  computed: {
    ...mapGetters('userNotifications', { 'notifications': 'unacknowledged' })
  },
  watch: {
    '$store.getters.currentUser': function () {
      this.fetchNotifications()
    }
  },
  beforeMount () {
    this.fetchNotifications()
    this.fetchInterval = setInterval(this.fetchNotifications, this.interval)
  },
  methods: {
    headline (notification) {
      let parts = []
      if (notification.created_at) parts.push(notification.created_at)
      if (notification.event) parts.push(notification.event)
      return parts.join(' â€¢ ')
    },

    fetchNotifications () {
      if (!this.interval || !this.$store.getters.currentUser || !this.$store.getters.currentUser.id) return 'cannot fetch'
      this.$store.dispatch('userNotifications/setFilters', {
        acknowledged: 0
      })
      this.$store.dispatch('userNotifications/getItems', {
        params: {
          userId: this.$store.getters.currentUser.id
        }
      })
    },

    dismiss (notification) {
      this.dismissMultiple([notification])
    },
    dismissAll () {
      this.dismissMultiple(this.notifications)
    },
    dismissMultiple (notification) {
      for (let i in notification) {
        this.$store.dispatch('userNotifications/acknowledgeItem', {
          params: {
            userId: this.$store.getters.currentUser.id,
            id: notification[i].id
          }
        })
      }
    }
  }
}
</script>

<style scoped>
  .notification-headline, .notification-msg {
    font-size: 12px;
  }
  .notification-headline {
    overflow: initial;
    white-space: initial;
  }
  .notification-msg {
    overflow: initial;
    white-space: initial;
  }

  .notification-label {
    height: auto;
    overflow: initial;
    white-space: initial;
    font-size: 14px;
  }

  /deep/ .v-list {
    padding: 0;
  }

  /deep/ .v-list__tile {
    height: auto;
    padding: 12px;
    align-items: start;
  }
  /deep/ .clear-tile .v-list__tile {
    padding: 2px 12px;
    align-items: center;
  }

  /deep/ .v-list__tile__avatar {
    min-width: auto;
    padding-right: 10px;
  }

  /deep/ .v-list__tile__action {
    min-width: auto ;
    padding-left: 12px;
  }
  /deep/ .v-list__tile__action i {
    font-size: 14px;
  }
</style>
