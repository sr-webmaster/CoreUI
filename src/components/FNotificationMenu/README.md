This was ported over but has some non-conventional aspects, so please treat this like a beta Component.

This component can change without any need for semver `major` version change.
