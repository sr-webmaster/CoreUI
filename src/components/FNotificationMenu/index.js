import Component from './FNotificationMenu'
export default Component
