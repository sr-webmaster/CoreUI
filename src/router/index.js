import VueRouter from 'vue-router'
import uniqBy from 'lodash/uniqBy'
import { makePageRoutes, standardizePageClass, getMeta } from './routes'

// Creates a function which runs the default next() callback and triggers subsequent Middleware function.
const nextFactory = function (context, middleware, index) {
  // Get next action
  let subsequentMiddleware = middleware[index]
  if (!subsequentMiddleware) return context.next

  return (...parameters) => {
    // Run the default Vue Router `next()` callback first.
    context.next(...parameters)
    // Then run the subsequent Middleware with a new `nextMiddleware()` callback.
    let nextMiddleware = nextFactory(context, middleware, index + 1)
    subsequentMiddleware({ ...context, next: nextMiddleware })
  }
}

/**
 * See README.md section on Pages
 */
const createMiddlewareGuard = (router) => {
  return (to, from, next) => {
    if (to.meta.middleware) {
      let middleware = Array.isArray(to.meta.middleware) ? to.meta.middleware : [to.meta.middleware]
      let context = {
        from,
        next,
        router,
        to
      }
      let nextMiddleware = nextFactory(context, middleware, 1)
      return middleware[0]({ ...context, next: nextMiddleware })
    }
    return next()
  }
}

const optionDefaults = {
  excludePage: false,
  redirectOnNotFound: true
}
export const install = (Vue, options = {}) => {
  Vue.use(VueRouter)
  options = { ...optionDefaults, ...options }
  // Get any page routes that have been passed in
  let primaryPageRoutes = []
  if (options.pages) {
    primaryPageRoutes = Array.isArray(options.pages) ? options.pages : makePageRoutes(options.pages, options)
  }

  /* Our Vue Router Object */
  const router = new VueRouter({
    /* Use Pretty URL */
    mode: 'history',
    /* Save The Scroll Position , Useful When Redirecting Back */
    scrollBehavior (to, from, savedPosition) {
      if (savedPosition) {
        return savedPosition
      } else {
        return { x: 0, y: 0 }
      }
    },
    routes: [
      // No route match, go to 404 page
      (options.redirectOnNotFound || !options.NotFoundPage)
        ? { path: '*', redirect: '/404' }
        : { path: '*', component: standardizePageClass(options.NotFoundPage), meta: getMeta(options.NotFoundPage, options.middleware) }
    ].concat(uniqBy(primaryPageRoutes, 'path'))
  })

  router.beforeEach(createMiddlewareGuard(router))

  Vue.router = router
  return router
}

export default (Vue, options) => {
  return install(Vue, options)
}
