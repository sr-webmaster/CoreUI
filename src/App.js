import mapValues from 'lodash/mapValues'
import get from 'lodash/get'
import { install as installVuetify } from './vuetify'
import { install as installRouter } from './router'
import { makePageRoutes } from './router/routes'
import { install as installAnalytics } from './analytics/google'
import { install as installStore, createStore } from './store'
import { install as installApiClient } from './http/client'
import { addInterceptor as authInterceptor } from './http/authHandlers'
import { addInterceptor as errorInterceptor } from './http/errorHandlers'
import { version as coreVersion } from '../package.json'
import themeDefault from './theme'

export const _createStoreModules = (providers, initialState = {}, options = {}) => {
  let modules = {}
  providers.forEach((provider) => {
    modules = {
      ...modules,
      ...mapValues(provider.store, (item, key) => {
        const value = item(initialState[key] || {}, { axios: options.axios })
        return {
          ...value,
          namespaced: true
        }
      })
    }
  })
  return modules
}

export const createStoreFromProviders = (providers, initialState = {}, options = {}) => {
  options.modules = _createStoreModules(providers, initialState, options)
  return createStore(initialState, options)
}

/**
 * Let's make it feel like a Vue App starting up
 */
class App {
  constructor (props = {}) {
    const excludePageDefault = (key) => {
      return key.indexOf('/register') > -1 && get(props, 'store.state.registrationType', null) === 2
    }

    this.Vue = props.Vue
    this.axios = props.axios
    this.analyticsOptions = {
      ...props.analytics
    }
    this._pendingPages = []
    this.routerMiddleware = props.middleware
    this.excludePage = props.excludePage || excludePageDefault
    this.register(props)
  }

  register (props = {}) {
    installStore(this.Vue)
    installVuetify(this.Vue, { theme: props.theme || themeDefault, ignoreVuetifyStyles: props.ignoreVuetifyStyles })
    installApiClient(this.Vue, { axios: this.axios })
    this.registerModules(props.modules)
    this.store = this.createStore(props.initialState)
    this._router = this.createRouter(props)
  }

  boot (AppComponent, options = {}) {
    installAnalytics(this.Vue, {
      id: this.analyticsOptions.id,
      router: this._router,
      errorMethod: console.error
    })
    const propComponents = options.components || {}
    this.vm = new this.Vue({
      router: this._router,
      el: '#app',
      components: {
        App: AppComponent,
        ...propComponents
      },
      template: '<App/>',
      store: this.store
    })
    authInterceptor(this.vm)
    errorInterceptor(this.vm)
  }

  createRouter (options = {}) {
    const router = installRouter(this.Vue, {
      redirectOnNotFound: options.redirectOnNotFound,
      NotFoundPage: options.NotFoundPage,
      middleware: this.routerMiddleware,
      pages: this._pendingPages.reverse(),
      excludePage: this.excludePage
    })
    this._pendingPages = []
    return router
  }

  createStore (initialState) {
    const options = {
      axios: this.axios
    }
    return createStoreFromProviders(this.modules, initialState, options)
  }

  registerModule (module) {
    const instance = module()
    this.modules.push(instance)
    if (instance.layouts) {
      this.addLayouts(instance.layouts)
    }
    if (instance.pages) {
      this.addPages(instance.pages)
    }
  }

  registerModules (list, options) {
    this.modules = Array.isArray(this.modules) ? this.modules : []
    if (!Array.isArray(list)) {
      throw new Error('registerModules must be called with an array of modules')
    }
    list.forEach((module) => {
      this.registerModule(module, {
        axios: this.axios,
        ...options
      })
    })
  }

  addLayouts (context) {
    context.keys().forEach(key => {
      this.Vue.component('layout-' + (key.slice(2, -4)), context(key).default)
    })
  }

  addPages (directory) {
    const options = {
      excludePage: this.excludePage,
      middleware: this.routerMiddleware
    }
    const pages = makePageRoutes(directory, options)
    if (this.getRouter()) {
      this.addRoutes(pages)
    } else {
      this._pendingPages = this._pendingPages.concat(pages.reverse())
    }
  }

  addRoutes (routes) {
    return this.getRouter().addRoutes(routes)
  }

  /**
   * Right now this may seem absurd but allows us to have a central interface for getting the Router that isn't the static property on Vue
   * @returns {*}
   */
  getRouter () {
    return this._router
  }

  getVersions () {
    let versions = {
      core: coreVersion
    }
    this.modules.forEach(module => {
      if (module.name) {
        versions[module.name] = module.version
      } else {
        console.error('unnamed FreshPlatform Module')
      }
    })
    return versions
  }
}

export default App
