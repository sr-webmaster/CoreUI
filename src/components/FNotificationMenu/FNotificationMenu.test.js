import { FIXTURE_USER } from '@freshinup/core-ui/tests/__data__/user'
import { mount, shallowMount } from '@vue/test-utils'
import { createStoreFromProviders } from '@freshinup/core-ui/src/App'
import Provider from '@freshinup/core-ui/src/Provider'
import createLocalVue from 'vue-cli-plugin-freshinup-ui/utils/testing/createLocalVue'
import Component from '@freshinup/core-ui/src/components/FNotificationMenu'
import { defaultStory } from './FNotificationMenu.stories'

describe('FNotificationMenu', () => {
  describe('Computed', () => {
    // None
  })

  describe('Methods', () => {
    let localVue, mock
    beforeEach(() => {
      const vue = createLocalVue()
      localVue = vue.localVue
      mock = vue.mock
    })
    afterEach(() => {
      mock.restore()
    })
    describe('fetchNotifications()', () => {
      test('interval (default) will fetch notifications', () => {
        mock
          .onGet('api/currentUser').reply(200, FIXTURE_USER)
          .onAny()
          .reply(config => {
            console.warn('No mock match for ' + config.url, config)
            return [404, {}]
          })
        const store = createStoreFromProviders([ Provider() ], {
          currentUser: FIXTURE_USER
        })
        store.dispatch = jest.fn()
        const wrapper = shallowMount(Component, {
          localVue,
          store
        })
        const result = wrapper.vm.fetchNotifications()
        expect(result).toBeUndefined()
        expect(store.dispatch).toHaveBeenCalledWith('userNotifications/getItems', {
          params: {
            userId: 1
          }
        })
      })
      test('interval set to falsey will not fetch notifications', () => {
        mock
          .onGet('api/currentUser').reply(200, FIXTURE_USER)
          .onAny()
          .reply(config => {
            console.warn('No mock match for ' + config.url, config)
            return [404, {}]
          })
        const store = createStoreFromProviders([ Provider() ], {
          currentUser: FIXTURE_USER
        })
        store.dispatch = jest.fn()
        const wrapper = shallowMount(Component, {
          propsData: {
            interval: 0
          },
          localVue,
          store
        })
        const result = wrapper.vm.fetchNotifications()
        expect(result).not.toBeUndefined()
        expect(result).toEqual('cannot fetch')
        expect(store.dispatch).not.toHaveBeenCalledWith('userNotifications/getItems')
      })
    })
  })

  describe('Visuals', () => {
    test('defaults', () => {
      // Use mount and snapshot assertions
      const wrapper = mount(defaultStory(), {
        localVue: createLocalVue().vue
      })

      expect(wrapper.isVueInstance()).toBe(true)
      expect(wrapper.element).toMatchSnapshot()
    })
  })
})

describe('NavigationMenu', () => {

})
