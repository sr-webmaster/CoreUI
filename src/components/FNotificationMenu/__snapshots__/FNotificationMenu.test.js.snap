// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FNotificationMenu Visuals defaults 1`] = `
<div
  class="container grid-list-xl"
>
  <div
    class="layout column"
  >
    <div
      class="flex"
    >
      <div
        class="v-alert warning"
      >
        <i
          aria-hidden="true"
          class="v-icon material-icons theme--light v-alert__icon"
        >
          priority_high
        </i>
        <div>
          
                  This component is currently in Beta. As such it can change without warning or semver Major version updates
                
        </div>
      </div>
    </div>
     
    <div
      class="flex"
    >
      <div
        class="v-card v-sheet theme--light"
        style="width: 320px;"
      >
        <div
          class="v-list theme--light"
          role="list"
        />
      </div>
       
      <button
        class="v-btn theme--light"
        type="button"
      >
        <div
          class="v-btn__content"
        >
          Reset
        </div>
      </button>
    </div>
  </div>
</div>
`;
